File: 6_4_PasportRF/src/pkg6_4_pasportrf/Pasport.java
package pkg6_4_pasportrf;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;
import java.util.TreeSet;
import pkg6_4_pasportrf.Main;

//Построить класс пользователь — Фамилия, Имя, Отчество, Дата рождения,
//телеграм. Реализовать вывод на экран, возможность сравнения по дате Рождения,
//возможность точного равенства, реализовать возможность добавления элементов в
//множество. Построить HashSet TreeSet с элементами этого класса, проверить корректность
//работы поиска элементов.

class Pasport implements Comparable<Pasport> {
    private String lastName;
    private String firstName;
    private String middleName;
    private Integer passportNumber;
    private Integer passportSeries;
    private LocalDate birthDate;
    
    private String Gender;
    private String PlaceOfbirth;
    private String WhoGaveItAway;
    private LocalDate DateOfIssue;
    private Integer UnitNumber;
    
//    public void setPassportNumber(String passportNumber) {
//        if (passportNumber != null && passportNumber.matches("\\d{4}")) {
//            this.passportNumber = passportNumber;
//        } else {
//            throw new IllegalArgumentException("Номер паспорта должен состоять из 4 цифр");
//        }
//    }
    
    public Pasport(String lastName, String firstName, String middleName, Integer passportNumber, Integer passportSeries, LocalDate birthDate, String Gender, String PlaceOfbirth, String WhoGaveItAway, LocalDate DateOfIssue, Integer UnitNumber) {   
        this.lastName = lastName; 
        this.firstName = firstName; 
        this.middleName = middleName; 
        this.passportNumber = passportNumber; 
        this.passportSeries = passportSeries;
        this.birthDate = birthDate;
        
        this.Gender = Gender;
        this.PlaceOfbirth = PlaceOfbirth;
        this.WhoGaveItAway = WhoGaveItAway;
        this.DateOfIssue = DateOfIssue;
        this.UnitNumber = UnitNumber;
    }

    @Override
    public String toString() {
        return "Pasport{" +
                "lastName='" + lastName + '\'' +
                ", firstName='" + firstName + '\'' +
                ", middleName='" + middleName + '\'' +
                ", passportNumber='" + passportNumber + '\'' +
                ", passportSeries='" + passportSeries + '\'' +
                ", birthDate=" + birthDate +
                
                ", Gender=" + Gender +
                ", PlaceOfbirth=" + PlaceOfbirth +
                ", WhoGaveItAway=" + WhoGaveItAway +
                ", DateOfIssue=" + DateOfIssue +
                ", UnitNumber=" + UnitNumber +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pasport pasport = (Pasport) o;
        return lastName.equals(pasport.lastName) &&
                firstName.equals(pasport.firstName) &&
                middleName.equals(pasport.middleName) &&
                passportNumber.equals(pasport.passportNumber) &&
                passportSeries.equals(pasport.passportSeries) &&
                birthDate.equals(pasport.birthDate) &&
                Gender.equals(pasport.Gender) &&
                PlaceOfbirth.equals(pasport.PlaceOfbirth) &&
                WhoGaveItAway.equals(pasport.WhoGaveItAway) &&
                DateOfIssue.equals(pasport.DateOfIssue) &&
                UnitNumber.equals(pasport.UnitNumber);
    }

    @Override
    public int hashCode() {
        int result = lastName.hashCode();
        result = 31 * result + firstName.hashCode();
        result = 31 * result + middleName.hashCode();
        result = 31 * result + passportNumber.hashCode();
        result = 31 * result + passportSeries.hashCode();
        result = 31 * result + birthDate.hashCode();
        result = 31 * result + Gender.hashCode();
        result = 31 * result + PlaceOfbirth.hashCode();
        result = 31 * result + WhoGaveItAway.hashCode();
        result = 31 * result + DateOfIssue.hashCode();
        result = 31 * result + UnitNumber.hashCode();
        return result;
    }

    @Override
    public int compareTo(Pasport other) {
        return this.birthDate.compareTo(other.birthDate);
    }
}

File: src/main/kotlin/Pasport.kt
import java.util.Date

// Задание 4
//Построить класс пользователь — Фамилия, Имя, Отчество, Дата рождения,
//телеграм. Реализовать вывод на экран, возможность сравнения по дате Рождения,
//возможность точного равенства, реализовать возможность добавления элементов в
//множество. Построить HashSet TreeSet с элементами этого класса, проверить корректность
//работы lkjhgпоиска элементов.

class Pasport(var lastname : String, var name : String, var fathername : String = "",
              var pasportNumber : Int, var pasportSeries : Int, var birth : Date,
              var gender : String, var placeOfBirth : String, var whoGaveItAway : String,
              var DateOfIssue : Date, var unitNumber : Int)  : Comparable<Pasport> {

    fun write() {
        println(
            "Пользователь\n" +
                    "Фамилия : $lastname\n" +
                    "Имя : $name\n" +
                    "Отчество : $fathername\n" +
                    "Номер паспорта : $pasportNumber\n" +
                    "Серия паспорта : $pasportSeries\n" +
                    "Дата рождения: ${birth.day}.${birth.month}.${birth.year}\n" +
                    "Пол : $gender\n" +
                    "Место рождения : $placeOfBirth\n" +
                    "Паспорт выдан : $whoGaveItAway\n" +
                    "Дата выдачи : $DateOfIssue\n" +
                    "Код подразделения : $unitNumber\n")
    }

    override fun equals(other: Any?): Boolean = if (other is Pasport) {
        (name == other.name) &&
                (lastname == other.lastname) &&
                (fathername == other.fathername) &&
                (pasportNumber == other.pasportNumber) &&
                (pasportSeries == other.pasportSeries) &&
                (birth == other.birth) &&
                (gender == other.gender) &&
                (placeOfBirth == other.placeOfBirth) &&
                (whoGaveItAway == other.whoGaveItAway) &&
                (DateOfIssue == other.DateOfIssue) &&
                (unitNumber == other.unitNumber)
    } else false

    override fun compareTo(other: Pasport): Int {
        return compareValuesBy(
            this, other,
            Pasport::birth, Pasport::lastname, Pasport::name, Pasport::fathername, Pasport::pasportNumber, Pasport::pasportSeries, Pasport::placeOfBirth, Pasport::whoGaveItAway, Pasport::DateOfIssue, Pasport::DateOfIssue, Pasport::unitNumber
        )
    }

    override fun hashCode(): Int {
        var result = birth.hashCode()
        result = 31 * result + lastname.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + fathername.hashCode()
        result = 31 * result + pasportNumber.hashCode()
        result = 31 * result + pasportSeries.hashCode()
        result = 31 * result + gender.hashCode()
        result = 31 * result + placeOfBirth.hashCode()
        result = 31 * result + whoGaveItAway.hashCode()
        result = 31 * result + DateOfIssue.hashCode()
        result = 31 * result + unitNumber.hashCode()

        return result
    }
}